packages/block-editor/src/store/reducer.js

```diff
@@ -192,6 +195,154 @@ export function isUpdatingSameBlockAttribute( action, lastAction ) {
 	);
 }
 
+/**
+ * Higher-order reducer intended to reset the cache key of all blocks
+ * whenever the post meta values change.
+ *
+ * @param {Function} reducer Original reducer function.
+ *
+ * @return {Function} Enhanced reducer function.
+ */
+const withPostMetaUpdateCacheReset = ( reducer ) => ( state, action ) => {
+	const newState = reducer( state, action );
+	const previousMetaValues = get( state, [ 'settings', '__experimentalMetaSource', 'value' ] );
+	const nextMetaValues = get( newState.settings.__experimentalMetaSource, [ 'value' ] );
+	// If post meta values change, reset the cache key for all blocks
+	if ( previousMetaValues !== nextMetaValues ) {
+		newState.blocks = {
+			...newState.blocks,
+			cache: mapValues( newState.blocks.cache, () => ( {} ) ),
+		};
+	}
+
+	return newState;
+};
+
+/**
+ * Higher-order reducer intended to compute a cache key for each block in the post.
+ * A new instance of the cache key (empty object) is created each time the block object
+ * needs to be refreshed (for any change in the block or its children).
+ *
+ * @param {Function} reducer Original reducer function.
+ *
+ * @return {Function} Enhanced reducer function.
+ */
+const withBlockCache = ( reducer ) => ( state = {}, action ) => {
+	const newState = reducer( state, action );
+
+	if ( newState === state ) {
+		return state;
+	}
+	newState.cache = state.cache ? state.cache : {};
+
+	const addParentBlocks = ( clientIds ) => {
+		return clientIds.reduce( ( result, clientId ) => {
+			let current = clientId;
+			do {
+				result.push( current );
+				current = state.parents[ current ];
+			} while ( current );
+			return result;
+		}, [] );
+	};
+
+	const fillKeysWithEmptyObject = ( clientIds ) => {
+		return clientIds.reduce( ( result, key ) => {
+			result[ key ] = {};
```

Comment:
We use empty object references as the key, would it be possible to be a simple integer that we increment on each change? The advantage being that it makes debugging easier, it would allow us to check if the key changed using the redux devtools or something similar, while right now cache always seems to contain the same valuel

Reply:
actually, the object works better because `rememo` relies on the fact that the dependencies are not scalar values.